pkg/helm_resources_manager/resources_monitor.go

```diff
@@ -128,46 +127,139 @@ func (r *ResourcesMonitor) Resume() {
+			// log.Errorf("%s: Cannot get GroupVersionResource info for apiVersion '%s' kind '%s' from api-server. Possibly CRD is not created before informers are started. Error was: %v", ei.Monitor.Metadata.DebugName, ei.Monitor.ApiVersion, ei.Monitor.Kind, err)
 			return nil, err
 		}
-		//log.Debugf("%s: GVR for kind '%s' is '%s'", ei.Monitor.Metadata.DebugName, ei.Monitor.Kind, ei.GroupVersionResource.String())
+		// log.Debugf("%s: GVR for kind '%s' is '%s'", ei.Monitor.Metadata.DebugName, ei.Monitor.Kind, ei.GroupVersionResource.String())
 
 		gvr := schema.GroupVersionResource{
 			Group:    apiRes.Group,
 			Version:  apiRes.Version,
 			Resource: apiRes.Name,
 		}
-		// Resources are filtered by metadata.name field. Object is considered absent if list is empty.
-		listOptions := v1.ListOptions{
-			FieldSelector: fields.OneTermEqualSelector("metadata.name", m.Name()).String(),
+		ns := m.Namespace(r.defaultNamespace)
+		if !apiRes.Namespaced {
+			ns = ""
+		}
+		nsgvr := namespacedGVR{
+			Namespace: ns,
+			GVR:       gvr,
 		}
 
-		var objList *unstructured.UnstructuredList
-
-		if apiRes.Namespaced {
-			ns := m.Namespace(r.defaultNamespace)
-			objList, err = r.kubeClient.Dynamic().Resource(gvr).Namespace(ns).List(context.TODO(), listOptions)
+		ex, ok := gvrMap[nsgvr]
+		if ok {
+			ex = append(ex, m)
+			gvrMap[nsgvr] = ex
 		} else {
-			objList, err = r.kubeClient.Dynamic().Resource(gvr).List(context.TODO(), listOptions)
+			gvrMap[nsgvr] = []manifest.Manifest{m}
 		}
-		if err != nil {
-			return nil, fmt.Errorf("Fetch list for helm resource %s: %s", m.Id(), err)
+	}
+
+	return gvrMap, nil
+}
+
+type gvrManifestResult struct {
+	hasAbsent bool
+	manifest  manifest.Manifest
+	err       error
+}
+
+func (r *ResourcesMonitor) checkGVRManifests(ctx context.Context, wg *sync.WaitGroup, nsgvr namespacedGVR, manifests []manifest.Manifest, resC chan gvrManifestResult, concurrency chan struct{}) {
+	defer wg.Done()
+
+	concurrency <- struct{}{}
+	defer func() {
+		<-concurrency
+	}()
+
+	// if context is cancelled - return
+	if ctx.Err() != nil {
+		return
+	}
+
+	existingObjs, err := r.listResources(ctx, nsgvr)
+	if err != nil {
+		resC <- gvrManifestResult{
+			err: err,
 		}
+		return
+	}
 
-		if len(objList.Items) == 0 {
-			res = append(res, m)
+	for _, mf := range manifests {
+		if _, ok := existingObjs[mf.Name()]; !ok {
+			resC <- gvrManifestResult{
+				hasAbsent: true,
+				manifest:  mf,
+			}
+			return
 		}
 	}
+}
+
+// list all objects in ns and return names of all existent objects
+func (r *ResourcesMonitor) listResources(ctx context.Context, nsgvr namespacedGVR) (map[string]struct{}, error) {
+	objList, err := r.kubeClient.Dynamic().Resource(nsgvr.GVR).Namespace(nsgvr.Namespace).List(ctx, v1.ListOptions{})
```

Comment:
if one list request fails, the whole upgrade checking fails too. Do we need to retry failed list requests in some cases?